### PR TITLE
MDL-72076 bulk actions ui: Implement bulk actions in qbank API

### PR DESCRIPTION
**Pull request by safatshahin**

### CODE REVIEW

File: question/bank/bulkmove/move.php [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+if ($category) {
+    list($tocategoryid, $contextid) = explode(',', $category);
+    if (! $tocategory = $DB->get_record('question_categories',
+        ['id' => $tocategoryid, 'contextid' => $contextid])) {
+        throw new \moodle_exception('cannotfindcate', 'question');
+    }
+}
+
+if ($moveselected && ($confirm = optional_param('confirm', '', PARAM_ALPHANUM))) {
+    $moveselected = required_param('moveselected', PARAM_RAW);
```
Comment (mattporritt) [inline question/bank/bulkmove/move.php:70]: move this one also

File: question/bank/bulkmove/move.php [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    if ($confirm == md5($moveselected)) {
+        if ($questionids = explode(',', $moveselected)) {
+            list($usql, $params) = $DB->get_in_or_equal($questionids);
+            $questions = $DB->get_records_sql("
```
Comment (mattporritt) [inline question/bank/bulkmove/move.php:74]: minor sql style formatting fixes required here.

File: question/bank/bulkmove/move.php [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+                        WHERE q.id {$usql}", $params);
+            foreach ($questions as $question) {
+                question_require_capability_on($question, 'move');
+            }
+            question_move_questions_to_category($questionids, $tocategory->id);
+        }
+    }
+    redirect(new moodle_url($returnurl, ['category' => "{$tocategoryid},{$contextid}"]));
+}
+
```
Comment (mattporritt) [inline question/bank/bulkmove/move.php:87]: remove extra line break

File: question/bank/bulkmove/move.php [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+if ($category) {
+    list($tocategoryid, $contextid) = explode(',', $category);
+    if (! $tocategory = $DB->get_record('question_categories',
+        ['id' => $tocategoryid, 'contextid' => $contextid])) {
+        throw new \moodle_exception('cannotfindcate', 'question');
+    }
+}
+
+if ($moveselected && ($confirm = optional_param('confirm', '', PARAM_ALPHANUM))) {
```
[the remark this answers is not in the document]
Comment (mattporritt) [reply question/bank/bulkmove/move.php:69]: This whole code block feels like it should be in a class/helper function as opposed to being embedded in the move page

File: question/bank/bulkmove/move.php [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+            question_move_questions_to_category($questionids, $tocategory->id);
+        }
+    }
+    redirect(new moodle_url($returnurl, ['category' => "{$tocategoryid},{$contextid}"]));
+}
+
+
+echo $OUTPUT->header();
+
+if ($moveselected) {
```
Comment (mattporritt) [inline question/bank/bulkmove/move.php:91]: This also should have some of the logic refactored into a helper method.

File: question/bank/bulkmove/move.php [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    $rawquestions = $_REQUEST;
```
Comment (mattporritt) [inline question/bank/bulkmove/move.php:92]: I'm not sure we should be accessing the super global directly, but I can't find any docs to say we can't

File: question/bank/bulkmove/tests/behat/bulk_move.feature [hunk 7]
```diff
@@ -0,0 +1,34 @@
+@qbank @qbank_bulkmove
```
Comment (mattporritt) [inline question/bank/bulkmove/tests/behat/bulk_move.feature:1]: Will need a behat test to cover actually moving something in bulk

File: question/bank/deletequestion/tests/behat/delete_question_column.feature [hunk 8]
```diff
@@ -0,0 +1,50 @@
+@qbank @qbank_deletequestion
```
Comment (mattporritt) [inline question/bank/deletequestion/tests/behat/delete_question_column.feature:1]: will need coverage to capture deleting

**Review by mattporritt — CHANGES_REQUESTED** [review]

File: question/bank/bulkmove/classes/bulk_move_action.php [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+/**
+ * Class bulk_move_action is the base class for moving questions.
+ *
+ * @package    qbank_bulkmove
+ * @copyright  2021 Catalyst IT Australia Pty Ltd
+ * @author     Safat Shahin <safatshahin@catalyst-au.net>
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class bulk_move_action extends \core_question\local\bank\bulk_action_base {
```
[the remark this answers is not in the document]
Comment (safatshahin) [reply question/bank/bulkmove/classes/bulk_move_action.php:27]: overridden methods, no docs for them

File: question/bank/bulkmove/classes/plugin_feature.php [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+/**
+ * Class plugin_feature is the entrypoint for the features.
+ *
+ * @package    qbank_bulkmove
+ * @copyright  2021 Catalyst IT Australia Pty Ltd
+ * @author     Safat Shahin <safatshahin@catalyst-au.net>
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class plugin_feature extends plugin_features_base {
+    public function get_bulk_actions(): ?object {
```
[the remark this answers is not in the document]
Comment (safatshahin) [reply question/bank/bulkmove/classes/plugin_feature.php:30]: overridden methods, no docs for them

File: question/bank/bulkmove/move.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (safatshahin) [reply question/bank/bulkmove/move.php:69]: moved to helper

File: question/bank/bulkmove/move.php (hunk 1 above)
Comment (safatshahin) [reply question/bank/bulkmove/move.php:70]: moved

File: question/bank/bulkmove/move.php (hunk 2 above)
Comment (safatshahin) [reply question/bank/bulkmove/move.php:74]: done

File: question/bank/bulkmove/tests/behat/bulk_move.feature (hunk 7 above)
Comment (safatshahin) [reply question/bank/bulkmove/tests/behat/bulk_move.feature:1]: covered by using the api

File: question/bank/deletequestion/tests/behat/delete_question_column.feature (hunk 8 above)
Comment (safatshahin) [reply question/bank/deletequestion/tests/behat/delete_question_column.feature:1]: done

**Comment by safatshahin** [comment]
sent to tracker review

**Pull request closed without merging**